cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileFetcher.java

```diff
@@ -0,0 +1,106 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.k8s.runtime;
+
+import io.cdap.cdap.common.BadRequestException;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.http.DefaultHttpRequestConfig;
+import io.cdap.cdap.common.internal.remote.RemoteClient;
+import io.cdap.common.http.HttpContentConsumer;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpRequest;
+import io.cdap.common.http.HttpRequests;
+import io.cdap.common.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.apache.twill.filesystem.Location;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.URI;
+import java.nio.ByteBuffer;
+
+/**
+ * Download file from AppFabric via internal REST API calls.
+ * <p>
+ * The target file could be either residing in AppFabric's local file system or a distributed file system
+ * that AppFabric is configured to access.
+ */
+class FileFetcher {
+  private static final Logger LOG = LoggerFactory.getLogger(FileFetcher.class);
+  private RemoteClient remoteClient;
+
+  FileFetcher(DiscoveryServiceClient discoveryClient) {
+    this.remoteClient = new RemoteClient(discoveryClient,
+                                         Constants.Service.APP_FABRIC_HTTP,
+                                         new DefaultHttpRequestConfig(false),
+                                         Constants.Gateway.INTERNAL_API_VERSION_3);
+  }
+
+  /**
+   * Download a file from AppFabric and store it in the target file.
+   *
+   * @param sourceURI uri to identity the file to download. This URI should exist in AppFabric.
+   * @param targetLocation target location to store the downloaded file
+   * @throws IOException if file downloading or writing to target location fails.
+   */
+  void download(URI sourceURI, Location targetLocation) throws IOException, BadRequestException {
+    OutputStream outputStream = targetLocation.getOutputStream();
+    HttpRequest request = remoteClient.requestBuilder(
+      HttpMethod.GET,
+      String.format("location/%s", sourceURI.getPath())).withContentConsumer(
+      new HttpContentConsumer() {
+        @Override
+        public boolean onReceived(ByteBuffer chunk) {
+          try {
+            byte[] bytes = new byte[chunk.remaining()];
+            chunk.get(bytes, 0, bytes.length);
+            outputStream.write(bytes);
+            outputStream.flush();
```

Comment:
Why need to flush?

Reply:
Flush so inmemory buffer can be freed (hoptefully), as an optimization.

Reply:
The buffers are rarely freed on flush, pretty often those are ringbuffers or reusable buffers. But flush may require a drive sync or network call

Reply:
if you want some optimization, your best bet is to check `chunk.hasArray` and use it's array if available. This would save you a buffer allocation and one copy

Reply:
Or actually best way is to use channels, as seen here: https://stackoverflow.com/questions/579600/how-to-put-the-content-of-a-bytebuffer-into-an-outputstream
```
public void writeBuffer(ByteBuffer buffer, OutputStream stream) {
   WritableByteChannel channel = Channels.newChannel(stream);

   channel.write(buffer);
}```

Reply:
Which in-memory buffer you are talking about? Internal to the FileOutputStream or the `bytes` you have? In either case, calling flush doesn't free anything. Depending on the implementation, flush at the best is to call `fsync` to force data to get written out to the storage device. However, memory usage is still the same after the flush because memory is not freed, but rather emptied. In this particular case, since we only use the file after it is closed, flushing on every write would only make the download slower, without any benefits. (flush/sync is useful when doing write-ahead-log that type of logic to prevent data loss on crashing).

Reply:
Original intention is to flush so the new byte[] can be freed right away. But it does depend on implementation. 
The concern I had was high memory usage due to too many new bytes[] calls across multiple iterations that underlying library holds onto them. 

Anyway, removed the flush